Add default delete action to actions dropdown in object detail modal

kamilmielnik opened the pull request:
Closes #32323

### Description

- Tests will be added in #32350

### How to verify

> Clicking delete should open a confirmation modal where the user has to confirm that they wish to delete the record and that there’s no undo.

> If a successful delete is run from an object detail modal, the detail modal should close and the user should see a toast. The row that was deleted should no longer be present in the table.

### Design notes
- The object detail modal has up/down arrows to navigate between neighboring rows. We don't need to close the object detail modal after deleting the object - we can navigate to next or previous object, or close the modal if there are no more objects. I believe [this was the original intention](https://www.notion.so/metabase/Allow-updates-and-deletes-on-model-object-detail-views-f05b298c3aba4a6ca883b2738d86824d?pvs=4#1ff1b9c44cd544d5b60881f80bcf272f) for the descoped part with objects displayed in dashboards. This PR implements closing both modals when deleting objects. If you want to try out how it would have worked with auto-navigation to sibling row please `git cherry-pick 8b24685eee5cd968f53cc14b7194a66588fb5845` into this branch
- Deleting an object that is referenced by other objects will often result in an error ("Referential integrity constraint violation") - see the end of demo - we might want to check if it's possible to hide/disable deleting in that case beforehand. Or consider cascade delete option?

### Demo

This demo uses 2 simple models that do nothing but reference sample db tables: accounts & reviews respectively.

https://github.com/metabase/metabase/assets/6830683/6b954457-1aa1-41c7-b154-26ffd05baaca

kamilmielnik committed:
Rename showActions to showControls

kamilmielnik committed:
Use Flex

kamilmielnik committed:
Import Flex from local modules

kamilmielnik committed:
Make horizontalAttachments a prop in EntityMenu

kamilmielnik committed:
Use EntityMenu to create actions dropdown

kamilmielnik committed:
Render actions menu conditionally

kamilmielnik committed:
Add useActionListQuery

kamilmielnik committed:
Fetch actions in ObjectDetailHeader

kamilmielnik committed:
Remove unused canRunActions prop

kamilmielnik committed:
Move logic to root component

kamilmielnik committed:
Fix tests

kamilmielnik committed:
Merge branch '32320-object-detail-actions' into 32321-actions-dropdown

kamilmielnik committed:
Use canRunAction and fetch databases in ObjectDetail view

kamilmielnik committed:
Rename actions to actionItems and getActions to getActionItems

kamilmielnik committed:
Extract getActionItems to utils

kamilmielnik committed:
Update tests

kamilmielnik committed:
Rename modelActions to actions

kamilmielnik committed:
Extract isValidImplicitDeleteAction and isValidImplicitUpdateAction out of getActionItems

kamilmielnik committed:
Consolidate getActionItems

kamilmielnik committed:
Use theme scale values for Mantine Flex component

kamilmielnik committed:
Either render all controls or none

kamilmielnik committed:
Fix default props/state in EntityMenu

kamilmielnik committed:
Fix props

kamilmielnik committed:
Update ObjectDetail unit tests

kamilmielnik committed:
Add unit tests for isValidImplicitDeleteAction and isValidImplicitUpdateAction

kamilmielnik committed:
Update tests with more specific non-implicit action cases

kamilmielnik committed:
Use singular form to match function interface

kamilmielnik committed:
Add unit tests for getActionItems

kamilmielnik committed:
Update test case name

kamilmielnik committed:
Sort imports

kamilmielnik committed:
Rename initialValues to values

kamilmielnik committed:
Rename dashcardParamValues to initialValues

kamilmielnik committed:
Lift fetchInitialValues up from ActionParametersInputForm

kamilmielnik committed:
Rename refetchValues to prefetchValues

kamilmielnik committed:
Fix tests

kamilmielnik committed:
Refactor shouldPrefetch to be a prop

kamilmielnik committed:
Get rid of dashboard and dashcard props from ActionParametersInputForm

kamilmielnik committed:
Merge branch '32320-object-detail-actions' into 32322-update-action

kamilmielnik committed:
Merge branch '32320-object-detail-actions' into 32322-update-action

kamilmielnik committed:
Lift the canPrefetch condition up as it is not desired outside of dashboards

kamilmielnik committed:
Hook update action with temporary API

kamilmielnik committed:
Use executeActionId in fetchInitialValues

kamilmielnik committed:
Fix post-merge conflict

kamilmielnik committed:
Rename prefetchValues to prefetchDashcardValues

kamilmielnik committed:
Use the new endpoint

kamilmielnik committed:
Merge branch '32320-object-detail-actions' into 32322-update-action

kamilmielnik committed:
Use rem values instead of scale values
- See https://github.com/metabase/metabase/pull/32331#discussion_r1267827835

kamilmielnik committed:
Merge branch '32320-object-detail-actions' into 32322-update-action

kamilmielnik committed:
Lift data reloading out of executeRowAction

kamilmielnik committed:
Add onSuccess prop to ActionExecuteModal

kamilmielnik committed:
Use onSubmitSuccess from ActionParametersInputForm

kamilmielnik committed:
Reload question after action is successful

kamilmielnik committed:
Fix stale prop issues in ActionParametersInputForm

kamilmielnik committed:
Fix object id in initialValues

kamilmielnik committed:
Add prefetchesInitialValues prop to make useActionForm more generic

kamilmielnik committed:
Ignore public actions for now

kamilmielnik committed:
Merge branch '32320-object-detail-actions' into 32322-update-action

kamilmielnik committed:
Fix id mixup

kamilmielnik committed:
Update unit test

kamilmielnik committed:
Add DeleteObjectModal

kamilmielnik committed:
Introduce useActionQuery

kamilmielnik committed:
Use useActionQuery in DeleteObjectModal

kamilmielnik committed:
Hook action execution

kamilmielnik committed:
Fix typo in copy

kamilmielnik committed:
Sort exports

kamilmielnik committed:
Remove redundant memoization

kamilmielnik committed:
Use modern JS syntax instead of noop

kamilmielnik committed:
Move getNextId to __support__/utils

kamilmielnik committed:
Extract isImplicitDeleteAction & isImplicitUpdateAction

kamilmielnik committed:
Reuse isImplicitDeleteAction

kamilmielnik committed:
Merge branch '32322-update-action' into 32323-delete-action

kamilmielnik committed:
Fix object modal getting closed with Escape key when delete or update modals are open

kamilmielnik committed:
Rename isUpdateModalOpen to isActionExecuteModalOpen, updateActionId to actionId

kamilmielnik committed:
Navigate to sibling row or close the modal on deletion

kamilmielnik committed:
Revert "Navigate to sibling row or close the modal on deletion"

This reverts commit 8b24685eee5cd968f53cc14b7194a66588fb5845.

kamilmielnik committed:
Close object detail on deletion

kamilmielnik committed:
Add error handling

kamilmielnik committed:
Add error handling

kamilmielnik committed:
Greatly simplify DeleteObjectModal
- remove useActionQuery usage
- remove useActionQuery
- remove DeleteObjectModal.styled.tsx

kamilmielnik committed:
Improve diff

kamilmielnik committed:
Shorten code

kdoh reviewed (commented):
Feeling good in general. A couple things are top of mind for me.

- I'm seeing [a bit of jumpiness](https://www.loom.com/share/a16687a78f9d4acf97803a0c8670ca63) on the update modal when the data loads. It's non blocking, but would be great to fix if it's a matter of when we do the content check.
- Conor and I will get you an answer on cascading deletes. We should determine if we want to block this PR on it or make a note in the Epic to follow up once we know what route we want to go. 
- R.e. the whole move to another record or close the object detail after a successful delete bit, I think closing the modal makes the most sense so let's continue on that route. Thanks for looking into that other approach, your intuition was right. Moving to another record felt more important in the dashboard context so there wasn't a content gap all of the sudden.

cdeweyx commented:
Looking real nice. Couple of open items before I approve:

> I'm seeing [a bit of jumpiness](https://www.loom.com/share/a16687a78f9d4acf97803a0c8670ca63) on the update modal when the data loads. It's non blocking, but would be great to fix if it's a matter of when we do the content check.

I would love to try and resolve this before merging. It's pretty in-your-face and while it's a split second, it makes me a little hesitant about the consequential thing I'm about to do.

> Conor and I will get you an answer on cascading deletes. We should determine if we want to block this PR on it or make a note in the Epic to follow up once we know what route we want to go.

I agree with Kyle's suggestion [here](https://metaboat.slack.com/archives/C057LE0Q4PQ/p1690295017296989?thread_ts=1690208614.530869&cid=C057LE0Q4PQ): "We’re able to tell beforehand that the current entity is referenced by other tables and add a “disabled” state to the delete option." That seems like the way to go for me. I don't think it has to block this PR but it's a definite follow-up item.

kamilmielnik committed:
Merge branch '32320-object-detail-actions' into 32323-delete-action

kamilmielnik committed:
Always cast objectId to number
- See https://metaboat.slack.com/archives/C057T1QTB3L/p1690292706101109?thread_ts=1690279562.697669&cid=C057T1QTB3L
  "the reason you’re not seeing it fail on the sample dataset is probably because you’re
  using h2 for sample-dataset. it’ll fail the same if you use a postgres one.
  h2 is probably more loose in type so it allows sending a string for id.
  nevertheless, I think you can use cast it to the correct type using the param type.
  I think we did that on the UI where we execute actions from model details page."

uladzimirdev commented on frontend/src/metabase/visualizations/components/ObjectDetail/DeleteObjectModal.tsx:
```diff
@@ -0,0 +1,70 @@
+}) => {
+  const dispatch = useDispatch();
+
+  const disabled =
```
so, am I right, disabled is only for submit button? can we rename it then?

uladzimirdev commented on frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.tsx:
```diff
@@ -410,18 +423,27 @@ export function ObjectDetailView({
         />
       </Modal>
+
+      <Modal isOpen={isDeleteModalOpen} onClose={handleDeleteModalClose}>
```
can't you move `Modal` wrapper into `DeleteObjectModal`?

kamilmielnik replied:
I wanted to be consistent with the current code:
- the `ActionExecuteModal` in lines above it
- there are lots (dozens) of components called `XYZModal` that render `<ModalContent>` as their root DOM element (examples: `CollectionMoveModal`, `AddToDashSelectDashModal`, `QuestionSavedModal`)

uladzimirdev replied:
it's not straightforward, but I agree with your points

kamilmielnik commented on frontend/src/metabase/visualizations/components/ObjectDetail/DeleteObjectModal.tsx:
```diff
@@ -0,0 +1,70 @@
+
+  const disabled =
+    typeof actionId === "undefined" ||
+    typeof objectId === "undefined" ||
```
I just dropped an [RFC about it](https://www.notion.so/metabase/56f6f59da3b647de89acb28ea9ebc424?v=cca33933b6244a7a94f256d1c110ce09&p=5e63b631c8c545a9bbd272e54deeeebe&pm=s).

kamilmielnik commented on frontend/src/metabase/visualizations/components/ObjectDetail/DeleteObjectModal.tsx:
```diff
@@ -0,0 +1,70 @@
+      await ActionsApi.execute({
+        id: actionId,
+        parameters: {
+          id: typeof objectId === "string" ? parseInt(objectId, 10) : objectId,
```
What's the advantage?

kamilmielnik commented on frontend/src/metabase/visualizations/components/ObjectDetail/DeleteObjectModal.tsx:
```diff
@@ -0,0 +1,70 @@
+  return (
+    <ModalContent
+      title={t`Are you sure you want to delete this row?`}
+      footer={[
```
It works differently, if it's an array there will automatically be margins between the buttons:
https://github.com/metabase/metabase/blob/3f35c0d96c8fab4d14bf3aa93464bfa80f801955/frontend/src/metabase/components/ModalContent/ModalContent.tsx#L191-L197

uladzimirdev replied:
oh, I didn't know that, thanks!

kamilmielnik committed:
Inline disabled into JSX

kamilmielnik committed:
Use a more friendly error message when failed to delete an object because it's referenced by other objects
- See: https://metaboat.slack.com/archives/C057LE0Q4PQ/p1690377421352469?thread_ts=1690208614.530869&cid=C057LE0Q4PQ

uladzimirdev commented:
LGTM
